File: mac/Voice/Hotkey/TapLatch.swift
import Foundation

/**
 Decides whether a key gesture is a hold, a double-tap that latches, or a stray tap.

 Pure, and deliberately clock-free: every method takes the time as a parameter and `Coordinator`
 supplies `Date()`. This is the same split already used for the release→paste measurement, where
 `pendingRelease`/`releaseAt` live on the Coordinator so `DictationMachine`'s effects stay
 assertable. A type that reads the clock itself can only be tested with sleeps, and a gesture test
 that sleeps is both slow and flaky.

 The trick that makes latching free is in `handle(.release)`. A press shorter than
 `DictationMachine.minimumMs` is *already* thrown away by the reducer as too short to be a
 dictation, so holding its stop open for a moment costs nothing that existed — while a real hold
 stops the instant the key comes up, with no window and no added latency.
 */
struct TapLatch {
    /// How long after a short release a second press still counts as a double-tap.
    static let windowMs = 300

    enum Outcome: Equatable {
        /// Begin recording. The dictation is real from this moment, even if it later turns out to
        /// have been a stray tap.
        case startDictation
        /// The key came up too quickly to be a dictation. Keep recording and start a timer for
        /// `TapLatch.windowMs`, calling `windowExpired(at:)` if it fires.
        ///
        /// Deliberately carries no deadline. An absolute `Date` in an `Equatable` payload cannot be
        /// asserted on: `t0 + 0.450` and `(t0 + 0.150) + 0.300` print identically and compare
        /// unequal. The caller knows `windowMs`, which is all it needs to schedule the timer.
        case holdOpen
        /// A second press landed inside the window. The recording that is already running becomes
        /// hands-free; nothing is restarted, so audio from the first tap is kept.
        case latch
        /// Stop recording and transcribe. Covers both a normal hold-release and the press that ends
        /// a latched session — from everything downstream they are the same event.
        case endSession
        /// The window closed with no second press. Discard; this was a stray key brush.
        case abandon
    }

    private enum State: Equatable {
        case idle
        case recording(pressedAt: Date)
        case windowOpen(releasedAt: Date)
        case latched
        /// The press that ended a latched session has been handled; its release must be swallowed
        /// so it does not read as the start of something new.
        case swallowingRelease
    }

    private var state: State = .idle

    /// Elapsed milliseconds, rounded to the granularity the decision is actually made at.
    ///
    /// `Date` arithmetic does not round-trip: `Date(1_000_000).addingTimeInterval(0.4)` minus
    /// `Date(1_000_000)` is 399.9999998 ms, not 400, so an exactly-400 ms press read as a tap and an
    /// exactly-300 ms double-tap could fail to latch. Both thresholds are expressed in whole
    /// milliseconds, so compare in whole milliseconds.
    private static func elapsedMs(from start: Date, to end: Date) -> Double {
        (end.timeIntervalSince(start) * 1000).rounded()
    }

    var isLatched: Bool { state == .latched }

    /// Whether a key event should still be interpreted at all. Used by the monitor to decide
    /// between the idle and listening event masks.
    var isActive: Bool { state != .idle }

    mutating func handle(_ action: HotkeyAction, at now: Date) -> [Outcome] {
        switch (state, action) {

        case (.idle, .press):
            state = .recording(pressedAt: now)
            return [.startDictation]

        case (.recording(let pressedAt), .release):
            let heldMs = Self.elapsedMs(from: pressedAt, to: now)
            // The boundary is `DictationMachine.minimumMs`, referenced rather than repeated:
            // "too short to be a dictation" and "short enough to be a tap" are the same threshold
            // and must never drift apart.
            if heldMs >= Double(DictationMachine.minimumMs) {
                state = .idle
                return [.endSession]
            }
            state = .windowOpen(releasedAt: now)
            return [.holdOpen]

        case (.windowOpen(let releasedAt), .press):
            if Self.elapsedMs(from: releasedAt, to: now) <= Double(Self.windowMs) {
                state = .latched
                return [.latch]
            }
            // Past the deadline. In practice the timer has already fired and moved us to `.idle`,
            // so this is only reachable when the window is driven manually — but it must still do
            // the right thing: bin the stray tap and treat this press as a fresh dictation.
            state = .recording(pressedAt: now)
            return [.abandon, .startDictation]

        case (.latched, .press):
            // Ends on key-down, not on release: waiting for the release would make stopping feel
            // laggy for no benefit.
            state = .swallowingRelease
            return [.endSession]

        // The release of the press that latched, and the release of the press that ended a latched
        // session. Both are physical key-ups with no meaning of their own.
        case (.latched, .release), (.swallowingRelease, .release):
            if case .swallowingRelease = state { state = .idle }
            return []

        case (_, .cancel):
            // Esc, or a shortcut chord during a hold. The dictation is discarded by the reducer;
            // all this type has to do is forget the gesture.
            state = .idle
            return []

        default:
            return []
        }
    }

    /// Called by `Coordinator` when the window timer fires.
    mutating func windowExpired(at now: Date) -> [Outcome] {
        guard case .windowOpen(let releasedAt) = state,
              Self.elapsedMs(from: releasedAt, to: now) >= Double(Self.windowMs) else { return [] }
        state = .idle
        return [.abandon]
    }
}

File: windows/Spit.Core/Hotkey/TapLatch.cs
namespace Spit.Core;

/// Port of mac/Voice/Hotkey/TapLatch.swift.
///
/// Decides whether a key gesture is a hold, a double-tap that latches, or a stray tap. Pure and
/// clock-free: every method takes the time as a parameter and the coordinator supplies it, so the
/// gesture tests run without sleeping.
///
/// Latching is free because of `Handle(Release)`: a press shorter than `DictationMachine.MinimumMs`
/// is already thrown away by the reducer as too short to be a dictation, so holding its stop open for
/// a moment costs nothing — while a real hold stops the instant the key comes up.
public sealed class TapLatch
{
    /// How long after a short release a second press still counts as a double-tap.
    public const int WindowMs = 300;

    public enum Outcome
    {
        /// Begin recording. The dictation is real from this moment, even if it later turns out to
        /// have been a stray tap.
        StartDictation,
        /// The key came up too quickly to be a dictation. Keep recording and start a timer for
        /// `TapLatch.WindowMs`, calling `WindowExpired` if it fires. Carries no deadline: the caller
        /// knows `WindowMs`, which is all it needs to schedule the timer.
        HoldOpen,
        /// A second press landed inside the window. The recording already running becomes
        /// hands-free; nothing is restarted, so audio from the first tap is kept.
        Latch,
        /// Stop recording and transcribe. Covers both a normal hold-release and the press that ends
        /// a latched session.
        EndSession,
        /// The window closed with no second press. Discard; this was a stray key brush.
        Abandon,
    }

    private abstract record State
    {
        public sealed record Idle : State;
        public sealed record Recording(DateTimeOffset PressedAt) : State;
        public sealed record WindowOpen(DateTimeOffset ReleasedAt) : State;
        public sealed record Latched : State;
        /// The press that ended a latched session has been handled; its release must be swallowed
        /// so it does not read as the start of something new.
        public sealed record SwallowingRelease : State;
    }

    private State state = new State.Idle();

    /// Elapsed milliseconds, rounded to the whole milliseconds both thresholds are expressed in (the
    /// Mac rounds because `Date` arithmetic does not round-trip; kept so the boundaries agree).
    private static double ElapsedMs(DateTimeOffset start, DateTimeOffset end) =>
        Math.Round((end - start).TotalMilliseconds, MidpointRounding.AwayFromZero);

    public bool IsLatched => state is State.Latched;

    /// Whether a key event should still be interpreted at all.
    public bool IsActive => state is not State.Idle;

    public IReadOnlyList<Outcome> Handle(HotkeyAction action, DateTimeOffset now)
    {
        switch (state, action)
        {
            case (State.Idle, HotkeyAction.Press):
                state = new State.Recording(now);
                return [Outcome.StartDictation];

            case (State.Recording r, HotkeyAction.Release):
                // The boundary is `DictationMachine.MinimumMs`, referenced rather than repeated: "too
                // short to be a dictation" and "short enough to be a tap" must never drift apart.
                if (ElapsedMs(r.PressedAt, now) >= DictationMachine.MinimumMs)
                {
                    state = new State.Idle();
                    return [Outcome.EndSession];
                }
                state = new State.WindowOpen(now);
                return [Outcome.HoldOpen];

            case (State.WindowOpen w, HotkeyAction.Press):
                if (ElapsedMs(w.ReleasedAt, now) <= WindowMs)
                {
                    state = new State.Latched();
                    return [Outcome.Latch];
                }
                // Past the deadline. Normally the timer has already fired; when driven manually, bin
                // the stray tap and treat this press as a fresh dictation.
                state = new State.Recording(now);
                return [Outcome.Abandon, Outcome.StartDictation];

            case (State.Latched, HotkeyAction.Press):
                // Ends on key-down, not on release: waiting for the release would make stopping feel laggy.
                state = new State.SwallowingRelease();
                return [Outcome.EndSession];

            // The release of the press that latched, and of the press that ended a latched session.
            case (State.Latched, HotkeyAction.Release):
                return [];
            case (State.SwallowingRelease, HotkeyAction.Release):
                state = new State.Idle();
                return [];

            case (_, HotkeyAction.Cancel):
                // Esc, or a shortcut chord during a hold. The reducer discards the dictation; this
                // type only has to forget the gesture.
                state = new State.Idle();
                return [];

            default:
                return [];
        }
    }

    /// Forgets the current gesture, for endings that do not come through a key event (the 90 s cap, a
    /// recorder that failed to start).
    public void Reset() => state = new State.Idle();

    /// Enters the latched state without a gesture, for a session started from the bar's mic button,
    /// so the key can end what the mouse started.
    public void ForceLatched() => state = new State.Latched();

    /// Called by the coordinator when the window timer fires.
    public IReadOnlyList<Outcome> WindowExpired(DateTimeOffset now)
    {
        if (state is not State.WindowOpen w || ElapsedMs(w.ReleasedAt, now) < WindowMs) return [];
        state = new State.Idle();
        return [Outcome.Abandon];
    }
}
